source/extensions/filters/http/lua/lua_filter.cc

```diff
@@ -428,6 +435,129 @@ int StreamHandleWrapper::luaLogCritical(lua_State* state) {
+  } else if (hash == "sha256") {
+    return EVP_sha256();
+  } else if (hash == "sha384") {
+    return EVP_sha384();
+  } else if (hash == "sha512") {
+    return EVP_sha512();
+  } else if (hash == "md5_sha1") {
+    return EVP_md5_sha1();
+  } else {
+    return nullptr;
+  }
+}
+
+int StreamHandleWrapper::luaVerifySignature(lua_State* state) {
+  // Step 1: get public key
+  auto ptr = lua_touserdata(state, 2);
+  auto key = reinterpret_cast<EVP_PKEY*>(ptr);
+
+  // Step 2: initialize EVP_MD_CTX
+  auto ctx = EVP_MD_CTX_new();
+
+  // Step 3: initialize EVP_MD
+  absl::string_view hash_func = luaL_checkstring(state, 3);
+  auto md = getDigest(hash_func);
+
+  if (md == nullptr) {
+    lua_pushboolean(state, false);
+    auto err_msg = absl::StrCat(hash_func, " is not supported.");
+    lua_pushlstring(state, err_msg.data(), err_msg.length());
+    EVP_MD_CTX_free(ctx);
+    return 2;
+  }
+
+  // Step 4: initialize EVP_DigestVerify
+  int ok = EVP_DigestVerifyInit(ctx, nullptr, md, nullptr, key);
+  if (!ok) {
+    lua_pushboolean(state, false);
+    absl::string_view err_msg = "Failed to initialize digest verify.";
+    lua_pushlstring(state, err_msg.data(), err_msg.length());
+    EVP_MD_CTX_free(ctx);
+    return 2;
+  }
+
+  // Step 5: verify signature
+  const char* signature = luaL_checkstring(state, 4);
+  int sigLen = luaL_checknumber(state, 5);
+  absl::string_view sig(signature, sigLen);
+
+  const char* clearText = luaL_checkstring(state, 6);
+  int textLen = luaL_checknumber(state, 7);
+  absl::string_view text(clearText, textLen);
+
+  ok = EVP_DigestVerify(ctx, reinterpret_cast<const uint8_t*>(sig.data()), sig.length(),
+                        reinterpret_cast<const uint8_t*>(text.data()), text.length());
+
+  // Step 6: check result
+  if (ok == 1) {
+    lua_pushboolean(state, true);
+    lua_pushnil(state);
+    EVP_MD_CTX_free(ctx);
+    return 2;
+  }
+
+  lua_pushboolean(state, false);
+  auto err_msg = absl::StrCat("Failed to verify digest. Error code: ", ok);
+  lua_pushlstring(state, err_msg.data(), err_msg.length());
+  EVP_MD_CTX_free(ctx);
+  return 2;
+}
+
+int StreamHandleWrapper::luaDecodeBase64(lua_State* state) {
+  // Get input string
+  absl::string_view str = luaL_checkstring(state, 2);
+  std::string output;
+
+  auto ok = absl::Base64Unescape(str, &output);
+
+  if (ok) {
+    lua_pushlstring(state, output.data(), output.length());
+  } else {
+    // If decode failed, push a nil into stack.
+    lua_pushnil(state);
+  }
+
+  return 1;
+}
+
+int StreamHandleWrapper::luaImportPublicKey(lua_State* state) {
+  // Get byte array the the length.
+  const char* str = luaL_checkstring(state, 2);
+  int n = luaL_checknumber(state, 3);
+
+  absl::string_view keyder(str, n);
+  CBS cbs({reinterpret_cast<const uint8_t*>(keyder.data()), keyder.length()});
+  auto key = EVP_parse_public_key(&cbs);
+  if (key == nullptr) {
+    lua_pushnil(state);
+  } else {
+    lua_pushlightuserdata(state, key);
```

Comment:
Is it possible to push `bssl::UniquePtr<EVP_PKEY>` so we don't need explicit release function?

Reply:
I didn't try but I don't think ```bssl::UniquePtr<EVP_PKEY>``` will work here. The requirement is that import public key once and use the public key to verify signature multiple times. That's why I separate the import from signature verification. If importing public key and verifying signature are in one function, ```bssl::UniquePtr<EVP_PKEY>``` will work perfectly.

Reply:
I mean't you store a pointer of `bssl::UniquePtr<EVP_PKEY>` in lua userdata, when Lua GC that object, the destructor of `bssl::UniquePtr<EVP_PKEY>` will be called so you don't need explicit release.

I think it should work even you import public key once and use the public key to verify signature multiple times, no?

Reply:
I tried to pass pointer of `bssl::UniquePtr<EVP_PKEY>` to Lua and the resource was not released. 

For simplicity, I used light userdata to store the pointer of `bssl::UniquePtr<EVP_PKEY>` and light userdata are not managed by the garbage collector.

Reply:
Yeah so my suggestion is use userdata by registering a wrapper type of EVP_PKEY... so the script author doesn't have to explicitly release resource.

Also with current structure, how to achieve "import public key once and use the public key to verify signature multiple times."? The tests doesn't do this, and import/release is tied to `StreamHandlerWrapper` (which is per request concept)

Reply:
I will add a wrapper for EVP_PKEY. 

In unit test `SignatureVerify`, public key was imported and then the key was used to verify for multiple times.

Reply:
Thanks @lizan! Add a wrapper for `EVP_PKEY` and `releasePublicKey` was removed. Very elegant solution.

Reply:
Why cannot tie import public key to `StreamHandlerWrapper`? For each request, "import public key once and use the public key to verify signature multiple times."

Reply:
Ah if you want do multiple times per request that's fine, I thought you want share the public key across multiple request.

Reply:
Yeah. Public key is not shared among requests. Could you please take a look? I think the current version should be good.